### PR TITLE
Design System: Improve tooltips position in RTL

### LINKED ISSUE #9598: Tooltips: Watch position while open
A while back @BrookeGraham fixed an issue that was newly exposed with dropdowns where they should switch positions while expanded because their position on the page changed and obscured the menu. That was here: #9060

We should do the same for tooltips. @miina pointed out here: https://github.com/google/web-stories-wp/issues/9460#issuecomment-956072696 that there's a few tooltips that are encountering the same problem.

Acceptance criteria: 
- If a tooltip, while open, changes position and is obscured by the end of the browser window or app space it should automatically move positions so that it is not obscured. Places this happens: `toggle consistent corner radius`

### PR DESCRIPTION
**Pull request by BrookeGraham**
## Context

Tooltip placement is normally repositioned _if_ the tip will be cutoff by the edge of the window. However, in the edge case where a tooltip is on a toggle that reveals a menu, the tooltip can cover newly shown buttons, lists, etc. As is the case with the `toggle consistent corner radius` tip on the corner radius toggle button. 

## Summary

When the tooltip toggle button is switched from one state to another, we are remounting the tooltip. This will reposition the tooltip's anchor. It will also close the tooltip even though the button, ie. tooltip anchor is still in focus. 

While digging into the tooltip used within the `radius` toggle button I stumbled across a couple issues. ~~First, the `Radius` component, as well and the `SizePosition` component were not using the `Tooltip` from `Design System`. By importing it directly, those components were not receiving the RTL mapping for `placement`.~~ [see comment thread](https://github.com/google/web-stories-wp/pull/9668#discussion_r746758391) Secondly,  in RTL mode the tooltips are cutoff on the left side of the screen. There is an offset where if a tooltip is cut off on the right edge of the window we offset the anchor point. This forces it to render up against the edge of the window. However, this doesn't apply in RTL. 

## Relevant Technical Choices

Since we use quite a few tooltips, tracking the anchor's location on the screen and repositioning it every time they change seemed a bit overkill for this edge case. Remounting seemed the smart choice.

Added a switch case in the if statement within `Tooltip`'s `positionPlacement` callback to solve the cutoff in RTL mode. If the popup would normally be rendered outside of the window, we are updating the placement to `{placement}-end`. 

### AFTER

https://user-images.githubusercontent.com/1820266/140964557-239adcd8-4dbe-455c-b7b7-bb128defc068.mov

https://user-images.githubusercontent.com/1820266/140964548-cbac9c6f-19dd-4a8c-b50b-7f6bb226b2e8.mov

https://user-images.githubusercontent.com/1820266/140964532-1f0edb79-98ae-489c-b37b-7dc450c2767a.mov

https://user-images.githubusercontent.com/1820266/140964543-1c3ac0e8-0177-4e5e-8930-c19c1118d4cc.mov

### BEFORE

https://user-images.githubusercontent.com/1820266/140965561-e4075235-1844-48f3-9b5c-206136a7c0c2.mov

https://user-images.githubusercontent.com/1820266/140965571-865783bd-ab8e-4b1e-b5bc-c5e49e60ed7d.mov

https://user-images.githubusercontent.com/1820266/140965578-e343cd56-49ab-4bf8-bc74-69908b490346.mov

## Testing Instructions

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1. Select an image on the canvas.
2. In the `Design` tab select the `Selection` dropdown / panel
3. Hover on the Lock Icon for `Radius` - you should see the tooltip 
4. Click on the lock icon. This should reveal/hide a menu with the different radius options. During reveal/hide the tooltip should no longer show until the next time you hover (or select icon via keyboard)
5. Check RTL mode!!
6. In RTL mode the same tooltip for the `radius` icon should no longer be cutoff by the edge of the screen. This goes for all tooltips on the far left ( found by hover on `radius` lock icon, `aspect ratio` lock icon, and `align bottom`) 
7. Lastly, check bottom cutoff still works in both RTL and LTR 
8. Open `Document` tab and go to `Background Audio` panel
9. Adjust window so that the trash can icon is just in view
10. Hover the trash can, tooltip should show above the icon, and not be cutoff the edge of the screen. 

### Does this PR have a security-related impact?
no

### Does this PR change what data or activity we track or use?
no

### Does this PR have a legal-related impact?
no

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #9598

### CODE REVIEW

File: packages/design-system/src/components/tooltip/index.js [hunk 1]
```diff
@@ -129,19 +129,35 @@ function Tooltip({
...
+        switch (currentPlacement) {
+          case currentPlacement.endsWith('start'):
+            setDynamicPlacement(currentPlacement.replace('-start', '-end'));
+            break;
+          case currentPlacement.endsWith('end'):
+            setDynamicPlacement(currentPlacement.replace('-end', '-start'));
+            break;
+          default:
+            setDynamicPlacement(`${currentPlacement}-start`);
+        }
```
Comment (BrittanyIRL) [inline packages/design-system/src/components/tooltip/index.js]: Just for readability, thoughts on pulling this out into its own function? I'm torn between thinking it'd help read or not, mostly it just feels weird to have a switch nested in an if statement.

File: packages/story-editor/src/components/panels/design/sizePosition/radius.js [hunk 2]
```diff
@@ -207,7 +206,10 @@ function RadiusControls({ selectedElements, pushUpdateForObject }) {
       </StackableGroup>
 
       <LockContainer>
-        <Tooltip title={__('Toggle consistent corner radius', 'web-stories')}>
+        <Tooltip
+          title={__('Toggle consistent corner radius', 'web-stories')}
+          key={borderRadius.locked}
```
Comment (BrittanyIRL) [inline packages/story-editor/src/components/panels/design/sizePosition/radius.js]: 🤔  if borderRadius.locked is a bool, should this maybe be something like 
```
`radius_locked_tooltip_${borderRadius.locked.toString()}`
``` 
so it still retriggers the tooltip render but doesn't pass a bool as a key?

**Review by BrittanyIRL — APPROVED** [review]
Looks awesome. Great notes and the solution makes sense and is really robust. 

This also fixes this issue: https://github.com/google/web-stories-wp/issues/6581!

File: packages/story-editor/src/components/panels/design/sizePosition/radius.js (hunk 2 above)
Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/sizePosition/radius.js]: GP!!

File: packages/design-system/src/components/tooltip/index.js (hunk 1 above)
Comment (BrookeGraham) [reply packages/design-system/src/components/tooltip/index.js]: Agreed. It is super weird. At one point I did pull it out, but ended up refactoring it back in, though I had a bit of other stuff going on then. I think I will pull it out and maybe add a little more comments since it is hard to follow.

File: packages/story-editor/src/components/panels/design/sizePosition/sizePosition.js [hunk 3]
```diff
@@ -49,7 +50,6 @@ import {
   inputContainerStyleOverride,
   useCommonObjectValue,
 } from '../../shared';
-import Tooltip from '../../../tooltip';
```
Comment (miina) [inline packages/story-editor/src/components/panels/design/sizePosition/sizePosition.js:52]: The editor was previously using a local `Tooltip` component which was based on the `Tooltip` from the design system, and was adjusting the tooltip placement based on if it was RTL or not (using the `TOOLTIP_RTL_PLACEMENT`).

I assume the `TOOLTIP_RTL_PLACEMENT` was not working fully as expected? Is the local `Tooltip` component obsolete now and can be fully replaced?

We're still using the local `Tooltip` in various places, if the RTL logic used in there was not working then perhaps it should be replaced with the design system `Tooltip` everywhere now? Thoughts?

Comment (BrittanyIRL) [reply packages/story-editor/src/components/panels/design/sizePosition/sizePosition.js:52]: I think this PR should mean that it can be entirely replaced. I'd missed that the Tooltip was localized to the editor this spring when we pulled it out into the DS (it was copied over so it would have been the same structure 🤔 ). Anyways, _yes_ let's replace it! 

I'll make a [new ticket](https://github.com/google/web-stories-wp/issues/9684), we should branch off of this and try replacing all of them there as proof of concept before this gets merged. 

@BrookeGraham would you mind taking this on?

Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/sizePosition/sizePosition.js:52]: Yes! I was just looking at this. I will double check that we can remove and dedupe.

Comment (BrittanyIRL) [reply packages/story-editor/src/components/panels/design/sizePosition/sizePosition.js:52]: thank you!

Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/sizePosition/sizePosition.js:52]: Ok. Turns out the local version of the tooltip both in `Story-Editor` and `Dashboard`, are wrapping  the Design System `Tooltip` and mapping to the RTL placement **correctly**. What isn't working is _my_ {placement}-end swap. So we don't need to create a new ticket to de-dupe. I just need to account for {placement}-end.  {placement} and {placement}-start are working as expected in both RTL and LTR.

Comment (BrookeGraham) [reply packages/story-editor/src/components/panels/design/sizePosition/sizePosition.js:52]: Alrighty, I am using the local version of Tooltip, thanks @miina. I have also fixed my inability to properly write a switch /case statement 🤦🏻‍♀️

File: packages/design-system/src/components/tooltip/index.js [hunk 4]
```diff
@@ -128,23 +128,58 @@ function Tooltip({
...
+  // We can sometimes render a tooltip too far to the left, ie. in RTL mode.
+  // when that is the case, we can switch to placement-start and the tooltip will no longer get cutoff.
+  const updatePlacement = useCallback(() => {
+    const currentPlacement = placementRef.current;
+    switch (currentPlacement) {
+      case PLACEMENT.BOTTOM_START:
+      case PLACEMENT.TOP_START:
+      case PLACEMENT.RIGHT_START:
+        // {placement}-START shouldn't ever appear in overflow so do nothing
+        break;
```
Comment (samwhale) [inline packages/design-system/src/components/tooltip/index.js:140]: good to be explicit. Leaving it as the default without a comment may be confusing.

File: packages/story-editor/src/components/panels/design/sizePosition/radius.js [hunk 5]
```diff
@@ -207,7 +207,10 @@ function RadiusControls({ selectedElements, pushUpdateForObject }) {
       </StackableGroup>
 
       <LockContainer>
-        <Tooltip title={__('Toggle consistent corner radius', 'web-stories')}>
+        <Tooltip
+          title={__('Toggle consistent corner radius', 'web-stories')}
+          key={borderRadius.locked.toString()}
```
Comment (samwhale) [inline packages/story-editor/src/components/panels/design/sizePosition/radius.js:214]: good pr description about this - super smart. Should we leave a comment here explaining what it does?

```js
// Remounts the component when `borderRadius.locked` changes
```

**Review by samwhale — APPROVED** [review]
excellante

File: packages/design-system/src/components/tooltip/index.js (hunk 4 above)
Comment (BrookeGraham) [reply packages/design-system/src/components/tooltip/index.js:140]: Yeh, I felt like if I didn't leave a comment others would be confused why they weren't accounted for.

**Review by maxyinger — APPROVED** [review]
Looks great. Thanks for the helpful comments!
